### PR TITLE
Restify PEP 307

### PR DESCRIPTION
**Pull request by serhiy-storchaka**

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
For #4.

File: pep-0307.txt [hunk 1]
```diff
@@ -5,767 +5,817 @@ Last-Modified: $Date$
...
-    example,
+Pickling new-style objects causes serious pickle bloat.  For
+example,
 
-        class C(object): # Omit "(object)" for classic class
-            pass
-        x = C()
-        x.foo = 42
-        print len(pickle.dumps(x, 1))
+.. code-block:: python
```
Comment (brettcannon) [inline pep-0307.txt]: I think this directive is the cause of the Travis failures. Updating Travis to install pygments wouldn't be hard, but do we know if python.org has it installed? @berkerpeksag , do you know?

Comment (Mariatta) [inline pep-0307.txt]: I think @ilevkivskyi's PR at pythondotorg repo is meant to address this https://github.com/python/pythondotorg/pull/1063

I haven't had chance to review / test it locally ..

Comment (ilevkivskyi) [inline pep-0307.txt]: IIRC, with pygments installed, docutils produces HTML that is correctly rendered by python.org without any changes to ``pep2html`` script (although still b/w). My PR adds little post-processing and CSS styles to actually make it colored on python.org

Comment (brettcannon) [inline pep-0307.txt]: So is it safe to update the Travis check to assume that pygments is installed on python.org for rendering the PEPs?

Comment (ilevkivskyi) [inline pep-0307.txt]: > So is it safe to update the Travis check to assume that pygments is installed on python.org for rendering the PEPs?

It is safe if pygments are actually installed on the machine were ``pep2html`` runs, I don't know if this is the case (my PR on ``pythondotorg`` adds pygments to ``requirements.txt`` but it is not merged yet).

**Comment by serhiy-storchaka** [comment]
If there are problems with this I can remove `code-block` directives.

**Comment by brettcannon** [comment]
@serhiy-storchaka why don't we remove them for now so this is no longer blocked; once Pygments is on www.python.org -- or we get the PEPs hosted on Read the Docs: python/core-workflow#5 -- we can add it back.

File: pep-0307.txt [hunk 2]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+API for specifying the pickling protocol to picklers.)
 
 
 Motivation
+==========
 
-    Pickling new-style objects causes serious pickle bloat.  For
-    example,
+Pickling new-style objects causes serious pickle bloat.  For
+example, ::
```
Comment (berkerpeksag) [inline pep-0307.txt]: I'd change this to "For example::".

File: pep-0307.txt [hunk 3]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+The same warning applies to previous Python versions, despite the
+presence of safety checks there.
+
+
+Extended ``__reduce__`` API
+===========================
+
+There are several APIs that a class can use to control pickling.
+Perhaps the most popular of these are ``__getstate__`` and
+``__setstate__; but the most powerful one is ``__reduce__``.  (There's
```
Comment (berkerpeksag) [inline pep-0307.txt]: missing \`\`

File: pep-0307.txt [hunk 4]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+also ``__getinitargs__``, and we're adding ``__getnewargs__`` below.)
+
+There are several ways to provide ``__reduce__`` functionality: a
+class can implement a ``__reduce__`` method or a ``__reduce_ex__`` method
+(see next section), or a reduce function can be declared in
+``copy_reg`` (``copy_reg.dispatch_table`` maps classes to functions).  The
+return values are interpreted exactly the same, though, and we'll
+refer to these collectively as ``__reduce__``.
+
+IMPORTANT: pickling of classic class instances does not look for a
```
Comment (berkerpeksag) [inline pep-0307.txt]: you can change this to \*\*Important:\*\* since we now have text formatting.

File: pep-0307.txt [hunk 5]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+
+It is sometimes useful to know the protocol version when
+implementing ``__reduce__``.  This can be done by implementing a
+method named ``__reduce_ex__`` instead of ``__reduce__``.  ``__reduce_ex__``,
+when it exists, is called in preference over ``__reduce__`` (you may
+still provide ``__reduce__`` for backwards compatibility).  The
+``__reduce_ex__`` method will be called with a single integer
+argument, the protocol version.
+
+The 'object' class implements both ``__reduce__`` and ``__reduce_ex__;
```
Comment (berkerpeksag) [inline pep-0307.txt]: missing \`\`

File: pep-0307.txt [hunk 6]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+implementation).
+
+If no ``__setstate__`` method exists, a default implementation is
+provided that assumes the state is a dictionary mapping instance
+variable names to values.  The default implementation tries two
+things:
+
+- First, it tries to call ``self.__dict__.update(state)``.
+
+- If the update() call fails with a RuntimeError exception, it
```
Comment (berkerpeksag) [inline pep-0307.txt]: You should probably wrap update(), RuntimeError, setattr(...) and (key, value) with backticks in order to be consistent with the rest of the file.

File: pep-0307.txt [hunk 7]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+The default ``__reduce__`` implementation will fail at pickling time
+for built-in types not mentioned above, and for new-style classes
+implemented in C:  if they want to be picklable, they must supply
+a custom ``__reduce__`` implementation under protocols 0 and 1.
+
+For new-style classes implemented in Python, the default
+``__reduce__`` implementation (``copy_reg._reduce``) works as follows:
+
+Let ``D`` be the class on the object to be pickled.  First, find the
+nearest base class that is implemented in ``C`` (either as a
```
Comment (berkerpeksag) [inline pep-0307.txt]: I think C here is the programming language, not a class named C.

File: pep-0307.txt [hunk 8]
```diff
@@ -5,767 +5,808 @@ Last-Modified: $Date$
...
+=====  ====  =====  =================================================
+  0       0      1  Reserved --- will never be used
+  1     127    127  Reserved for Python standard library
+128     191     64  Reserved for Zope
+192     239     48  Reserved for 3rd parties
+240     255     16  Reserved for private use (will never be assigned)
+256     MAX    MAX  Reserved for future assignment
+=====  ====  =====  =================================================
+
+MAX stands for 2147483647, or ``2**31-1``.  This is a hard limitation
```
Comment (berkerpeksag) [inline pep-0307.txt]: You can probably use backticks for constants like `MAX`.

**Pull request merged**